src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncExceptionHandlerRewriter.cs

```diff
@@ -1004,14 +1007,18 @@ private sealed class AwaitCatchFrame
             private readonly Dictionary<LocalSymbol, LocalSymbol> _hoistedLocals;
             private readonly List<LocalSymbol> _orderedHoistedLocals;
 
-            public AwaitCatchFrame(SyntheticBoundNodeFactory F, TryStatementSyntax tryStatementSyntax)
+            public AwaitCatchFrame(SyntheticBoundNodeFactory F, TryStatementSyntax tryStatementSyntax, AwaitCatchFrame parentOpt = null)
             {
                 this.pendingCaughtException = new SynthesizedLocal(F.CurrentFunction, TypeWithAnnotations.Create(F.SpecialType(SpecialType.System_Object)), SynthesizedLocalKind.TryAwaitPendingCaughtException, tryStatementSyntax);
                 this.pendingCatch = new SynthesizedLocal(F.CurrentFunction, TypeWithAnnotations.Create(F.SpecialType(SpecialType.System_Int32)), SynthesizedLocalKind.TryAwaitPendingCatch, tryStatementSyntax);
 
                 this.handlers = new List<BoundBlock>();
-                _hoistedLocals = new Dictionary<LocalSymbol, LocalSymbol>();
-                _orderedHoistedLocals = new List<LocalSymbol>();
+                _hoistedLocals = parentOpt != null
+                    ? new Dictionary<LocalSymbol, LocalSymbol>(parentOpt._hoistedLocals)
```

Comment:
>parentOpt._hoistedLocals [](http://example.com/codeflow?start=63&length=24)

Instead of copying data, could we simply chain  ```AwaitCatchFrame``` and iterate through the chain when appropriate?